lib/double_entry/account.rb

```diff
@@ -27,35 +29,53 @@ def account(identifier, options = {})
 
       # @api private
       def currency(identifier)
-        accounts.detect { |a| a.identifier == identifier }.try(:currency)
+        accounts.find_without_scope(identifier).try(:currency)
       end
     end
 
     # @api private
-    class Set < Array
+    class Set
+      extend Forwardable
+
+      delegate [:each, :map] => :all
+
       def define(attributes)
-        self << Account.new(attributes)
+        Account.new(attributes).tap do |account|
+          if find_without_scope(account.identifier)
+            fail DuplicateAccount
+          else
+            backing_collection[account.identifier] = account
+          end
+        end
       end
 
       def find(identifier, scoped)
-        found_account = detect do |account|
-          account.identifier == identifier && account.scoped? == scoped
-        end
-        fail UnknownAccount, "account: #{identifier} scoped?: #{scoped}" unless found_account
-        found_account
-      end
+        found_account = find_without_scope(identifier)
 
-      def <<(account)
-        if any? { |a| a.identifier == account.identifier }
-          fail DuplicateAccount
+        if found_account && found_account.scoped? == scoped
+          found_account
         else
-          super
+          fail UnknownAccount, "account: #{identifier} scoped?: #{scoped}"
         end
       end
 
+      def find_without_scope(identifier)
+        backing_collection[identifier]
+      end
+
       def active_record_scope_identifier(active_record_class)
         ActiveRecordScopeFactory.new(active_record_class).scope_identifier
       end
+
+      def all
+        backing_collection.values
```

Comment:
That's true! However this is declared as part of the private API only, and only used internally in the jackhammer testing script, so I figured ordering is probably not a worry.